### 实践三：调用方可处理或需要处理的异常，使用Checked Exception，导致程序错误的请使用Runtime Exception

Java编译器会对"Checked Exception"进行检查，而对"运行时异常"不会检查。

也就是说，对于被检查的异常，要么通过throws进行声明抛出，要么通过try-catch进行捕获处理，否则不能通过编译。

而对于运行时异常，倘若既"没有通过throws声明抛出它"，也"没有用try-catch语句捕获它"，还是会编译通过。

当然，虽说Java编译器不会检查运行时异常，但是，我们同样可以通过throws对该异常进行说明，或通过try-catch进行捕获。

推荐编写javadoc注释时，使用@throws列举出运行时异常。

对于运行时异常，我们应该通过修改代码进行避免它的产生。
比如以下代码：

``` java
     /**
       *bala bala 巴拉
       *@return 员工，可能为空
       *@throws IllegalArgumentException
       */
     public Employee findByUserCode(int userCode){
       if(userCode <=0 ){
         throw new IllegalArgumentException(“userCode:”+userCode);
       }
     }
```
调用方应该在调用之前检查参数，而不是将非法参数传递下去。

而对于被检查的异常，则可以通过处理让程序恢复运行。

比如以下代码，根据门店地址，调用百度api更新门店坐标，根据地址反查经纬度可能失败，如果失败，我们设置默认值。

``` java
	public void updateOrganizationCoordinate(Organization org){
	     try{
	         Coordinate latlng= orgAddressToCoordiante(org.getAddress(),org.getCity());
	         org.setLat(latlng.getLat());
	        org.setLng(latlng.getLng());
	     }catch(AddressNotFoundException e){
	       // log exception
	      
	        org.setLat(0);
	        org.setLng(0);
	     }
	     
	}	
	
	public Coordinate orgAddressToCoordinate(String orgAddress,String city) throws AddressNotFoundException {
	   /////
	}
	
``` 

随便提一下，字符集 `Charset.forName(“utf-8”)`会抛出待检查异常UnsupportedCharsetException，JDK1.7以后我们可以用：`StandardCharsets.UTF_8`

